blockchain.go

```diff
@@ -137,14 +149,77 @@ func ValidateSignatures(pubKeys []ecc.PublicKey, platformPubKey ecc.PublicKey) e
 	return fmt.Errorf("platform pub key not found in deposit txn")
 }
 
+func isTransfer(action *eos.Action) bool {
+	return action.Name == eos.ActN("transfer")
+}
+
+func isDepositBon(action *eos.Action) bool {
+	return action.Name == eos.ActN("depositbon")
+}
+
 func isNewGame(action *eos.Action) bool {
 	return action.Name == eos.ActN("newgame") || action.Name == eos.ActN("newgameaffl")
 }
 
+func isNewGameBon(action *eos.Action) bool {
+	return action.Name == eos.ActN("newgamebon")
+}
+
 func isGameAction(action *eos.Action) bool {
 	return action.Name == eos.ActN("gameaction")
 }
 
+func getInvariantName(action *eos.Action) (string, error) {
+	switch {
+	case isTransfer(action):
+		return "transfer", nil
+	case isDepositBon(action):
+		return "depositbon", nil
+	case isNewGame(action):
+		return "newgame", nil
+	case isNewGameBon(action):
+		return "newgamebon", nil
+	case isGameAction(action):
+		return "gameaction", nil
+	}
+	return "", fmt.Errorf("action is not allowed")
+}
+
+func extractInvariant(actions []*eos.Action) ([]string, error) {
+	var invariant []string
+	for _, act := range actions {
+		name, err := getInvariantName(act)
+		if err != nil {
+			return nil, err
+		}
+		invariant = append(invariant, name)
+	}
+	return invariant, nil
+}
+
+func isInvariantAllowed(invariant []string) bool {
+	for _, inv := range allowedInvariants {
+		if len(inv) != len(invariant) {
+			continue
+		}
+
+		isBreak := false
```

Comment:
i think `matches := true` would be a better name

Reply:
ok